File: android-app/app/src/main/java/com/feelbachelor/doompedia/data/update/PackUpdateService.kt
package com.feelbachelor.doompedia.data.update

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import com.feelbachelor.doompedia.data.importer.DeltaApplier
import com.feelbachelor.doompedia.data.importer.DownloadProgressSnapshot
import com.feelbachelor.doompedia.data.importer.PackInstaller
import com.feelbachelor.doompedia.data.importer.PackManifest
import com.feelbachelor.doompedia.data.importer.PackShard
import com.feelbachelor.doompedia.data.importer.ShardDownloader
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import java.io.File
import java.net.URL
import java.security.MessageDigest

enum class PackUpdateStatus {
    NO_MANIFEST,
    SKIPPED_NETWORK,
    UP_TO_DATE,
    UPDATED,
    FAILED,
}

data class PackUpdateResult(
    val status: PackUpdateStatus,
    val installedVersion: Int,
    val installedSignature: String,
    val message: String,
)

data class PackUpdateProgress(
    val phase: String,
    val percent: Float,
    val downloadedBytes: Long,
    val totalBytes: Long,
    val bytesPerSecond: Long,
    val detail: String = "",
)

class PackUpdateService(
    private val context: Context,
    private val downloader: ShardDownloader,
    private val packInstaller: PackInstaller,
    private val deltaApplier: DeltaApplier,
    private val json: Json = Json { ignoreUnknownKeys = true },
) {
    suspend fun checkAndApply(
        manifestUrl: String,
        wifiOnly: Boolean,
        installedVersion: Int,
        installedSignature: String,
        onProgress: ((PackUpdateProgress) -> Unit)? = null,
    ): PackUpdateResult {
        if (manifestUrl.isBlank()) {
            return PackUpdateResult(
                status = PackUpdateStatus.NO_MANIFEST,
                installedVersion = installedVersion,
                installedSignature = installedSignature,
                message = "Manifest URL is empty",
            )
        }

        return runCatching {
            if (wifiOnly && !isOnWifi()) {
                return PackUpdateResult(
                    status = PackUpdateStatus.SKIPPED_NETWORK,
                    installedVersion = installedVersion,
                    installedSignature = installedSignature,
                    message = "Wi-Fi only mode is enabled. Disable it in Settings to download over mobile data.",
                )
            }

            onProgress?.invoke(
                PackUpdateProgress(
                    phase = "Fetching manifest",
                    percent = 0f,
                    downloadedBytes = 0L,
                    totalBytes = 0L,
                    bytesPerSecond = 0L,
                    detail = manifestUrl,
                )
            )
            val manifest = fetchManifest(manifestUrl)
            val remoteSignature = manifest.signature()
            val isSameOrOlderVersion = manifest.version <= installedVersion
            if (isSameOrOlderVersion && remoteSignature == installedSignature) {
                return PackUpdateResult(
                    status = PackUpdateStatus.UP_TO_DATE,
                    installedVersion = installedVersion,
                    installedSignature = installedSignature,
                    message = "No new pack available",
                )
            }

            val updateRoot = File(context.filesDir, "doompedia/updates/${manifest.packId}/v${manifest.version}")
            updateRoot.mkdirs()

            val deltaApplied = tryApplyDelta(
                manifestUrl = manifestUrl,
                manifest = manifest,
                updateRoot = updateRoot,
                wifiOnly = wifiOnly,
                installedVersion = installedVersion,
                onProgress = onProgress,
            )

            if (!deltaApplied) {
                applyFullPack(
                    manifestUrl = manifestUrl,
                    manifest = manifest,
                    updateRoot = updateRoot,
                    wifiOnly = wifiOnly,
                    onProgress = onProgress,
                )
            }

            onProgress?.invoke(
                PackUpdateProgress(
                    phase = "Completed",
                    percent = 100f,
                    downloadedBytes = manifest.shards.sumOf { it.bytes },
                    totalBytes = manifest.shards.sumOf { it.bytes },
                    bytesPerSecond = 0L,
                    detail = "Installed pack v${manifest.version}",
                )
            )
            PackUpdateResult(
                status = PackUpdateStatus.UPDATED,
                installedVersion = manifest.version,
                installedSignature = remoteSignature,
                message = "Updated to pack version ${manifest.version}",
            )
        }.getOrElse { error ->
            PackUpdateResult(
                status = PackUpdateStatus.FAILED,
                installedVersion = installedVersion,
                installedSignature = installedSignature,
                message = error.message ?: "Unknown update failure",
            )
        }
    }

    private suspend fun fetchManifest(manifestUrl: String): PackManifest {
        return withContext(Dispatchers.IO) {
            val payload = URL(manifestUrl).openStream().bufferedReader().use { it.readText() }
            json.decodeFromString(PackManifest.serializer(), payload)
        }
    }

    private suspend fun tryApplyDelta(
        manifestUrl: String,
        manifest: PackManifest,
        updateRoot: File,
        wifiOnly: Boolean,
        installedVersion: Int,
        onProgress: ((PackUpdateProgress) -> Unit)? = null,
    ): Boolean {
        val delta = manifest.delta ?: return false
        if (delta.baseVersion != installedVersion) return false

        val deltaUrl = resolveUrl(manifestUrl, delta.url)
        val localDelta = File(updateRoot, delta.url.substringAfterLast('/'))
        onProgress?.invoke(
            PackUpdateProgress(
                phase = "Downloading delta",
                percent = 0f,
                downloadedBytes = 0L,
                totalBytes = 0L,
                bytesPerSecond = 0L,
                detail = localDelta.name,
            )
        )
        downloader.download(
            sourceUrl = deltaUrl,
            destination = localDelta,
            wifiOnly = wifiOnly,
            resume = true,
            onProgress = { snapshot ->
                onProgress?.invoke(
                    snapshot.toPackProgress(
                        phase = "Downloading delta",
                        detail = localDelta.name,
                    )
                )
            }
        )

        val digest = localDelta.sha256()
        if (!digest.equals(delta.sha256, ignoreCase = true)) {
            localDelta.delete()
            return false
        }

        deltaApplier.apply(localDelta)
        return true
    }

    private suspend fun applyFullPack(
        manifestUrl: String,
        manifest: PackManifest,
        updateRoot: File,
        wifiOnly: Boolean,
        onProgress: ((PackUpdateProgress) -> Unit)? = null,
    ) {
        val totalBytes = manifest.shards.sumOf { it.bytes }.coerceAtLeast(1L)
        var completedBytes = 0L

        val localManifest = manifest.copy(
            shards = manifest.shards.map { shard ->
                val localName = shard.url.substringAfterLast('/')
                val localFile = File(updateRoot, localName)
                downloader.download(
                    sourceUrl = resolveUrl(manifestUrl, shard.url),
                    destination = localFile,
                    wifiOnly = wifiOnly,
                    resume = true,
                    onProgress = { snapshot ->
                        val globalDownloaded = (completedBytes + snapshot.downloadedBytes).coerceAtMost(totalBytes)
                        val percent = ((globalDownloaded.toDouble() / totalBytes.toDouble()) * 100.0).toFloat()
                        onProgress?.invoke(
                            PackUpdateProgress(
                                phase = "Downloading shards",
                                percent = percent,
                                downloadedBytes = globalDownloaded,
                                totalBytes = totalBytes,
                                bytesPerSecond = snapshot.bytesPerSecond,
                                detail = shard.id,
                            )
                        )
                    }
                )
                val digest = localFile.sha256()
                require(digest.equals(shard.sha256, ignoreCase = true)) {
                    "Checksum mismatch for shard ${shard.id}"
                }
                completedBytes += localFile.length().coerceAtMost(shard.bytes)

                PackShard(
                    id = shard.id,
                    url = localName,
                    sha256 = shard.sha256,
                    records = shard.records,
                    bytes = localFile.length(),
                )
            }
        )

        onProgress?.invoke(
            PackUpdateProgress(
                phase = "Installing",
                percent = 100f,
                downloadedBytes = totalBytes,
                totalBytes = totalBytes,
                bytesPerSecond = 0L,
                detail = "Applying downloaded content",
            )
        )
        File(updateRoot, "manifest.json").writeText(
            json.encodeToString(PackManifest.serializer(), localManifest),
        )
        packInstaller.installFromDirectory(updateRoot, expectedPackId = manifest.packId)
    }

    private fun resolveUrl(manifestUrl: String, shardUrl: String): String {
        return runCatching {
            URL(URL(manifestUrl), shardUrl).toString()
        }.getOrElse {
            if (shardUrl.startsWith("http://") || shardUrl.startsWith("https://")) {
                shardUrl
            } else {
                "$manifestUrl/$shardUrl"
            }
        }
    }

    private fun isOnWifi(): Boolean {
        val manager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val network = manager.activeNetwork ?: return false
        val capabilities = manager.getNetworkCapabilities(network) ?: return false
        return capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)
    }
}

private fun PackManifest.signature(): String {
    val raw = buildString {
        append(packId).append('|')
        append(language).append('|')
        append(version).append('|')
        append(createdAt).append('|')
        append(recordCount).append('|')
        append(compression).append('|')
        append(shards.size).append('|')
        shards.forEach { shard ->
            append(shard.id).append(':')
            append(shard.records).append(':')
            append(shard.bytes).append(':')
            append(shard.sha256).append(';')
        }
        delta?.let {
            append("|delta:")
                .append(it.baseVersion).append(':')
                .append(it.targetVersion).append(':')
                .append(it.sha256)
        }
    }
    val digest = MessageDigest.getInstance("SHA-256")
    val bytes = digest.digest(raw.toByteArray(Charsets.UTF_8))
    return bytes.joinToString(separator = "") { "%02x".format(it) }
}

private fun DownloadProgressSnapshot.toPackProgress(
    phase: String,
    detail: String,
): PackUpdateProgress {
    val boundedTotal = totalBytes.coerceAtLeast(0L)
    val boundedDownloaded = downloadedBytes.coerceAtLeast(0L)
    val percent = if (boundedTotal > 0L) {
        ((boundedDownloaded.toDouble() / boundedTotal.toDouble()) * 100.0)
            .coerceIn(0.0, 100.0)
            .toFloat()
    } else {
        0f
    }
    return PackUpdateProgress(
        phase = phase,
        percent = percent,
        downloadedBytes = boundedDownloaded,
        totalBytes = boundedTotal,
        bytesPerSecond = bytesPerSecond.coerceAtLeast(0L),
        detail = detail,
    )
}

private fun File.sha256(): String {
    val digest = MessageDigest.getInstance("SHA-256")
    inputStream().use { stream ->
        val buffer = ByteArray(DEFAULT_BUFFER_SIZE)
        while (true) {
            val read = stream.read(buffer)
            if (read <= 0) break
            digest.update(buffer, 0, read)
        }
    }
    return digest.digest().joinToString(separator = "") { "%02x".format(it) }
}

File: ios-app/Doompedia/Data/PackUpdateService.swift
import CryptoKit
import Foundation

enum PackUpdateStatus {
    case noManifest
    case skippedNetwork
    case upToDate
    case updated
    case failed
}

struct PackUpdateResult {
    let status: PackUpdateStatus
    let installedVersion: Int
    let message: String
}

struct PackUpdateProgress {
    let phase: String
    let percent: Double
    let downloadedBytes: Int64
    let totalBytes: Int64
    let bytesPerSecond: Int64
    let detail: String
}

final class PackUpdateService {
    private let installer: PackInstaller
    private let deltaApplier: DeltaApplier
    private let downloader: ShardDownloader
    private let decoder = JSONDecoder()

    init(installer: PackInstaller, deltaApplier: DeltaApplier, downloader: ShardDownloader) {
        self.installer = installer
        self.deltaApplier = deltaApplier
        self.downloader = downloader
    }

    func checkAndApply(
        manifestURLString: String,
        wifiOnly: Bool,
        installedVersion: Int,
        onProgress: ((PackUpdateProgress) -> Void)? = nil
    ) async -> PackUpdateResult {
        if manifestURLString.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            return PackUpdateResult(status: .noManifest, installedVersion: installedVersion, message: "Manifest URL is empty")
        }

        do {
            guard let manifestURL = URL(string: manifestURLString) else {
                return PackUpdateResult(status: .failed, installedVersion: installedVersion, message: "Invalid manifest URL")
            }

            if wifiOnly && !NetworkMonitor.shared.isOnline {
                return PackUpdateResult(
                    status: .skippedNetwork,
                    installedVersion: installedVersion,
                    message: "Wi-Fi only mode is enabled"
                )
            }

            onProgress?(PackUpdateProgress(
                phase: "Fetching manifest",
                percent: 0,
                downloadedBytes: 0,
                totalBytes: 0,
                bytesPerSecond: 0,
                detail: manifestURLString
            ))
            let manifest = try await fetchManifest(url: manifestURL)
            if manifest.version <= installedVersion {
                return PackUpdateResult(status: .upToDate, installedVersion: installedVersion, message: "No new pack available")
            }

            if manifest.compression != "none" && manifest.compression != "gzip" {
                return PackUpdateResult(
                    status: .failed,
                    installedVersion: installedVersion,
                    message: "Unsupported pack compression: \(manifest.compression)"
                )
            }

            let updateRoot = try updateDirectory(packID: manifest.packId, version: manifest.version)
            let deltaApplied = try await tryApplyDelta(
                manifestURL: manifestURL,
                manifest: manifest,
                updateRoot: updateRoot,
                wifiOnly: wifiOnly,
                installedVersion: installedVersion,
                onProgress: onProgress
            )

            if !deltaApplied {
                try await applyFullPack(
                    manifestURL: manifestURL,
                    manifest: manifest,
                    updateRoot: updateRoot,
                    wifiOnly: wifiOnly,
                    onProgress: onProgress
                )
            }

            onProgress?(PackUpdateProgress(
                phase: "Completed",
                percent: 100,
                downloadedBytes: manifest.shards.reduce(0) { $0 + $1.bytes },
                totalBytes: manifest.shards.reduce(0) { $0 + $1.bytes },
                bytesPerSecond: 0,
                detail: "Installed pack v\(manifest.version)"
            ))
            return PackUpdateResult(
                status: .updated,
                installedVersion: manifest.version,
                message: "Updated to pack version \(manifest.version)"
            )
        } catch {
            return PackUpdateResult(
                status: .failed,
                installedVersion: installedVersion,
                message: error.localizedDescription
            )
        }
    }

    private func fetchManifest(url: URL) async throws -> PackManifest {
        let (data, response) = try await URLSession.shared.data(from: url)
        guard let http = response as? HTTPURLResponse,
              (200 ... 299).contains(http.statusCode) else {
            throw NSError(domain: "Doompedia", code: 71, userInfo: [NSLocalizedDescriptionKey: "Manifest request failed"])
        }

        return try decoder.decode(PackManifest.self, from: data)
    }

    private func tryApplyDelta(
        manifestURL: URL,
        manifest: PackManifest,
        updateRoot: URL,
        wifiOnly: Bool,
        installedVersion: Int,
        onProgress: ((PackUpdateProgress) -> Void)? = nil
    ) async throws -> Bool {
        guard let delta = manifest.delta,
              delta.baseVersion == installedVersion else {
            return false
        }

        let remoteURL = resolveURL(base: manifestURL, path: delta.url)
        let localURL = updateRoot.appendingPathComponent(URL(fileURLWithPath: delta.url).lastPathComponent)
        onProgress?(PackUpdateProgress(
            phase: "Downloading delta",
            percent: 0,
            downloadedBytes: 0,
            totalBytes: 0,
            bytesPerSecond: 0,
            detail: localURL.lastPathComponent
        ))
        try await downloader.download(
            sourceURL: remoteURL,
            destinationURL: localURL,
            wifiOnly: wifiOnly,
            resume: true
        ) { downloaded, total, speed in
            let percent = total > 0 ? (Double(downloaded) / Double(total)) * 100.0 : 0
            onProgress?(PackUpdateProgress(
                phase: "Downloading delta",
                percent: max(0, min(100, percent)),
                downloadedBytes: downloaded,
                totalBytes: total,
                bytesPerSecond: speed,
                detail: localURL.lastPathComponent
            ))
        }

        let digest = try sha256(url: localURL)
        guard digest.caseInsensitiveCompare(delta.sha256) == .orderedSame else {
            try? FileManager.default.removeItem(at: localURL)
            return false
        }

        _ = try deltaApplier.apply(from: localURL)
        return true
    }

    private func applyFullPack(
        manifestURL: URL,
        manifest: PackManifest,
        updateRoot: URL,
        wifiOnly: Bool,
        onProgress: ((PackUpdateProgress) -> Void)? = nil
    ) async throws {
        var localShards: [PackShard] = []
        let totalBytes = max(1, manifest.shards.reduce(Int64(0)) { $0 + $1.bytes })
        var completedBytes: Int64 = 0

        for shard in manifest.shards {
            let remoteURL = resolveURL(base: manifestURL, path: shard.url)
            let localName = URL(fileURLWithPath: shard.url).lastPathComponent
            let localURL = updateRoot.appendingPathComponent(localName)
            try await downloader.download(
                sourceURL: remoteURL,
                destinationURL: localURL,
                wifiOnly: wifiOnly,
                resume: true
            ) { downloaded, _, speed in
                let global = min(totalBytes, completedBytes + downloaded)
                let percent = (Double(global) / Double(totalBytes)) * 100.0
                onProgress?(PackUpdateProgress(
                    phase: "Downloading shards",
                    percent: max(0, min(100, percent)),
                    downloadedBytes: global,
                    totalBytes: totalBytes,
                    bytesPerSecond: speed,
                    detail: shard.id
                ))
            }

            let digest = try sha256(url: localURL)
            guard digest.caseInsensitiveCompare(shard.sha256) == .orderedSame else {
                throw NSError(
                    domain: "Doompedia",
                    code: 72,
                    userInfo: [NSLocalizedDescriptionKey: "Checksum mismatch for shard \(shard.id)"]
                )
            }
            completedBytes += min(shard.bytes, Int64((try? localURL.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? 0))

            localShards.append(
                PackShard(
                    id: shard.id,
                    url: localName,
                    sha256: shard.sha256,
                    records: shard.records,
                    bytes: Int64((try? localURL.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? 0)
                )
            )
        }

        let localManifest = PackManifest(
            packId: manifest.packId,
            language: manifest.language,
            version: manifest.version,
            createdAt: manifest.createdAt,
            recordCount: manifest.recordCount,
            description: manifest.description,
            packTags: manifest.packTags,
            compression: manifest.compression,
            shards: localShards,
            delta: manifest.delta,
            topicDistribution: manifest.topicDistribution,
            entityDistribution: manifest.entityDistribution,
            sampleKeywords: manifest.sampleKeywords,
            attribution: manifest.attribution
        )

        let localManifestURL = updateRoot.appendingPathComponent("manifest.json")
        let data = try JSONEncoder().encode(localManifest)
        try data.write(to: localManifestURL, options: .atomic)
        onProgress?(PackUpdateProgress(
            phase: "Installing",
            percent: 100,
            downloadedBytes: totalBytes,
            totalBytes: totalBytes,
            bytesPerSecond: 0,
            detail: "Applying downloaded content"
        ))
        _ = try installer.install(from: updateRoot, expectedPackID: manifest.packId)
    }

    private func updateDirectory(packID: String, version: Int) throws -> URL {
        let base = try FileManager.default.url(
            for: .applicationSupportDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        )
        let dir = base
            .appendingPathComponent("doompedia")
            .appendingPathComponent("updates")
            .appendingPathComponent(packID)
            .appendingPathComponent("v\(version)")
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    private func resolveURL(base: URL, path: String) -> URL {
        URL(string: path, relativeTo: base) ?? base.appendingPathComponent(path)
    }

    private func sha256(url: URL) throws -> String {
        let data = try Data(contentsOf: url)
        let digest = SHA256.hash(data: data)
        return digest.map { String(format: "%02x", $0) }.joined()
    }
}
